### PR TITLE
connector: add azure connector

### PR DESCRIPTION
**Pull request by summerwind**
As we discussed in #504, Azure AD contains a non-standard claim. This PR adds the Azure connector to support non-standard claims.

### CODE REVIEW

**Comment by ericchiang** [comment]
Thanks for this!

Part of adding a connector is adding additional documentation to `Documentation/connectors-configuration.md`. Can you do that? It also helps me test your change.

**Comment by summerwind** [comment]
Thanks for the comment! I added a document for the `azure` connector. 
To test the azure connector, requires your own Azure Active Directory.
https://azure.microsoft.com/en-us/services/active-directory/

File: connector/connector_azure.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+			log.Errorf("Unable to construct claims: %v", err)
+			q.Set("error", oauth2.ErrorUnsupportedResponseType)
+			q.Set("error_description", "unable to construct claims")
+			redirectError(w, errorURL, q)
+			return
+		}
+
+		// ID Token issued by Azure AD does not contain a email claim.
+		// So we need to copy the value from upn claim to email claim internally.
+		upn, ok, err := claims.StringClaim("upn")
```
Comment (ericchiang) [inline connector/connector_azure.go]: Hmm, it seems like this is the only thing unique to this from the OpenID Connect plugin. Could we share code a bit more between those two or even just add a flag to the existing one?

**Comment by ericchiang** [comment]
Great. I'll sign up for an account tomorrow.

Can you take a look at my inline comment? I dislike just copying so much code from the existing OpenID Connect connector. Would like to see that addressed.

**Comment by ericchiang** [comment]
@summerwind yes I would be in favor of a more minimal change. We can still provide azure specific docs like we do with flavors of LDAP[0]. I still have to read a bit into the azure implementation to give good advice on which route to take.

[0] https://github.com/coreos/dex/blob/master/Documentation/connectors-configuration.md#example-searching-a-freeipa-server-with-groups

**Comment by summerwind** [comment]
@ericchiang I agree to take the way to become a minimal change. Please reply if you have any good advice!

**Comment by summerwind** [comment]
Any updates? Shall I modify/improve it somehow?

**Comment by ericchiang** [comment]
Sorry, I'd love to see this connector share code with the existing OpenID Connect implementation. Probably through some sort of configuration flag. Does that sound reasonable?

**Comment by summerwind** [comment]
@ericchiang Sure! I will open another PR when ready.

**Comment by ericchiang** [comment]
Just update this one :)

**Comment by summerwind** [comment]
@ericchiang I added `emailClaim` setting to OIDC Connector as an alternative way. Could you review please?

**Comment by ericchiang** [comment]
The last commit looks really good. Could you clean up your commit history by removing the unused commits? If you have issues with git feel free to open another PR.

**Comment by summerwind** [comment]
@ericchiang Thanks for the review. I have cleaned up my commits and updated PR.
Please tell me if you have another concern ;)

**Comment by ericchiang** [comment]
lgtm

**Pull request merged**